src/Microsoft.ServiceFabric.Actors/Runtime/VolatileActorStateProvider.cs

```diff
@@ -1122,6 +1142,157 @@ private static string ExtractStateName(ActorId actorId, string storageKey)
             return storageKey.Substring(storageKeyPrefix.Length + 1);
         }
 
+        private async Task StartStateProviderInitializationAsync(CancellationToken cancellationToken)
+        {
+            Exception unexpectedException = null;
+
+            try
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                await this.actorStateProviderHelper.ExecuteWithRetriesAsync(
+                    async () =>
+                    {
+                        await this.InitializeAndStartLogicalTimeManagerAsync(cancellationToken);
+                    },
+                    "StartStateProviderInitializationAsync",
+                    cancellationToken);
+            }
+            catch (OperationCanceledException opEx)
+            {
+                if (!cancellationToken.IsCancellationRequested)
+                {
+                    unexpectedException = opEx;
+                }
+            }
+            catch (FabricObjectClosedException)
+            {
+                // This can happen when replica is closing. CancellationToken should get signaled.
+                // Fall through and let the task check for CancellationToken.
+            }
+            catch (FabricNotPrimaryException)
+            {
+                // This replica is no more primary. CancellationToken should get signaled.
+                // Fall through and let the task check for CancellationToken.
+            }
+            catch (Exception ex)
+            {
+                unexpectedException = ex;
+            }
+
+            if (unexpectedException != null)
+            {
+                var mssgFormat = "StartStateProviderInitializationAsync() failed due to " +
+                                 "an unexpected Exception causing replica to fault: {0}";
+
+                ActorTrace.Source.WriteErrorWithId(
+                    TraceType,
+                    this.traceId,
+                    string.Format(mssgFormat, unexpectedException.ToString()));
+
+                this.partition.ReportFault(FaultType.Transient);
+            }
+        }
+
+        private async Task CancelStateProviderInitializationAsync()
+        {
+            if (this.stateProviderInitCts != null &&
+                this.stateProviderInitCts.IsCancellationRequested == false)
+            {
+                ActorTrace.Source.WriteInfoWithId(TraceType, this.traceId, "Canceling state provider initialization...");
+
+                this.stateProviderInitCts.Cancel();
+
+                try
+                {
+                    await this.stateProviderInitTask;
+                }
+                catch (Exception ex)
+                {
+                    // Code should never come here.
+                    ReleaseAssert.Failfast(

```

Comment:
I advise to handle TaskCanceledException as well along with OperationCanceledException inside StartStateProviderInitializationAsync to ensure we are not asserting for cancelations

Reply:
TaskCanceledException inherits from the OperationCanceledException, we dont need to catch that separately.